src/canvas.cpp

```diff
@@ -28,153 +28,260 @@
-      m_background_color(background_color)
-{
-}
-
-canvas::~canvas()
-{
-  if(m_surface != nullptr) {
-    cairo_surface_destroy(m_surface);
-  }
-
-  if(m_context != nullptr) {
-    cairo_destroy(m_context);
-  }
-}
-
-int canvas::width() const
-{
-  return gtk_widget_get_allocated_width(m_drawing_area);
-}
-
-int canvas::height() const
-{
-  return gtk_widget_get_allocated_height(m_drawing_area);
-}
-
-void canvas::initialize(GtkWidget *drawing_area)
-{
-  g_return_if_fail(drawing_area != nullptr);
-
-  m_drawing_area = drawing_area;
-  m_surface = create_surface(m_drawing_area);
-  m_context = create_context(m_surface);
-  m_camera.update_widget(width(), height());
-
-  // Draw to the newly created surface for the first time.
-  redraw();
-
-  // Connect to configure events in case our widget changes shape.
-  g_signal_connect(m_drawing_area, "configure-event", G_CALLBACK(configure_event), this);
-  // Connect to draw events so that we draw our surface to the drawing area.
-  g_signal_connect(m_drawing_area, "draw", G_CALLBACK(draw_surface), this);
-
-  // GtkDrawingArea objects need specific events enabled explicitly.
-  gtk_widget_add_events(GTK_WIDGET(m_drawing_area), GDK_BUTTON_PRESS_MASK);
-  gtk_widget_add_events(GTK_WIDGET(m_drawing_area), GDK_BUTTON_RELEASE_MASK);
-  gtk_widget_add_events(GTK_WIDGET(m_drawing_area), GDK_POINTER_MOTION_MASK);
-  gtk_widget_add_events(GTK_WIDGET(m_drawing_area), GDK_SCROLL_MASK);
-
-  g_info("canvas::initialize successful.");
-}
-
-void canvas::redraw()
-{
-  // Clear the screen and set the background color
-  cairo_set_source_rgb(m_context, m_background_color.red / 255.0,
-		       m_background_color.green / 255.0, m_background_color.blue / 255.0);
-  cairo_paint(m_context);
-
-  using namespace std::placeholders;
-  renderer g(m_context, std::bind(&camera::world_to_screen, m_camera, _1), &m_camera, m_surface);
-  m_draw_callback(g);
-
-  gtk_widget_queue_draw(m_drawing_area);
-
-  g_info("The canvas will be redrawn.");
-}
-
-renderer canvas::create_temporary_renderer()
-{
-  using namespace std::placeholders;
-  renderer g(m_context, std::bind(&camera::world_to_screen, m_camera, _1), &m_camera, m_surface);
-
-  return g;
-}
+    static cairo_surface_t *create_surface(GtkWidget *widget)
+    {
+      GdkWindow *parent_window = gtk_widget_get_window(widget);
+      int const width = gtk_widget_get_allocated_width(widget);
+      int const height = gtk_widget_get_allocated_height(widget);
+
+      // Cairo image surfaces are more efficient than normal Cairo surfaces
+      // However, you cannot use X11 functions to draw on image surfaces
+    #ifdef EZGL_USE_X11
+      return gdk_window_create_similar_surface(parent_window, CAIRO_CONTENT_COLOR_ALPHA, width, height);
+    #else
+      return gdk_window_create_similar_image_surface(
+          parent_window, CAIRO_FORMAT_ARGB32, width, height, 0);
+    #endif
+    }
+
```

Comment:
The comment at the top of this file says that create_and_generate_pdf creates a pdf file. But this function seems instead just to create a surface, and it is in fact canvas::print_pdf() that does the full work of creating a surface and generating a pdf (partially by calling this function).  Which routine is the externally visible routine that ezgl users should call?

Reply:
EZGL users should only call the canvas::print_pdf(), I will make the comment more clear. I am wondering if the EZGL users wouldn't need to call the create_and_generate_pdf at all, do I ever need to add such many lines of comments to it? I wrote the comment for create_and_generate_pdf before I wrote the canvas::print_pdf() function, I am not sure if I should just delete them.

Reply:
You can shorten the comments if they're not externally visible, and definitely the comments should reflect what the functions do.  These functions don't do that much, so the comments can be shorter. You can also delete the functions if you don't really need them.  It seems like you could just inline the 2 lines of code they perform into the calling function, as I think each is only called in one place.  Then you could delete them (and if they are really only called from one place and have two lines of function body it would be better to delete them).

Reply:
Ok, I just removed these create_and_generate_X functions and directly call the cairo surface create functions inside the print_X functions.